Apache Spark support (#8784)
fishtown-analytics/docs.getdbt.com: website/docs/docs/local/connect-data-platform/spark-setup.md

## Fusion and Spark

<Constant name="fusion" /> uses the Databricks SQL dialect for [static analysis](/docs/fusion/new-concepts#principles-of-static-analysis) when working with Spark. Databricks SQL is a superset of Spark SQL, so your SQL is validated with Databricks semantics. This provides comprehensive error checking and SQL comprehension features. A dedicated Spark SQL dialect for static analysis is planned for a future release.

## Authentication

The Spark adapter in <Constant name="fusion" /> supports:

- Thrift
  - Simple Authentication and Security Layer (SASL) PLAIN
  - No SASL (NOSASL) 
- Livy
  - Basic authentication (username and password)
  - When deployed on Amazon Web Services (AWS): AWS Signature Version 4
    - Supports authentication using single sign-on, service accounts, or user tokens

## Configure Fusion

Configure your Spark connection in `profiles.yml`:

<File name='~/.dbt/profiles.yml'>

```yaml
your_profile_name:
  target: dev
  outputs:
    dev:
      type: spark

>       method: [thrift | http | livy]

Reviewer: ```suggestion
      method: THRIFT_OR_HTTP_OR_LIVY
```

>       port: [port number]

Reviewer: ```suggestion
      port: PORT_NUMBER
```

>       auth: [authentication method]

Reviewer: ```suggestion
      auth: AUTHENTICATION_METHOD
```

>       schema: [database/schema name]

Reviewer: ```suggestion
      schema: DATABASE_OR_SCHEMA_NAME
```

>       host: [yourorg.sparkhost.com]

Reviewer: ```suggestion
      host: SPARK_HOSTNAME
```

>       platform_hint: [aws_emr_serverless | aws_emr_eks]  # optional

Reviewer: ```suggestion
      platform_hint: AWS_EMR_SERVERLESS_OR_AWS_EMR_EKS  # optional
```

      server_side_parameters:  # optional; required keys depend on platform_hint

>         livy.server.session.ttl: [seconds]  # optional; when using method: livy

Reviewer: ```suggestion
        livy.server.session.ttl: SESSION_TTL_SECONDS  # optional; when using method: livy
```

Reviewer: Hiya @luna-bianca 

Following the style guide, should lines 55 to 62 be in caps / screaming snake case?

Reviewer 2: Hey @nataliefiann I initially copied the format used in other pages that are not using the recommended case -- changed to screaming snake case now. Thanks for letting me know!

```

</File>

| Profile field | Required | Description | Example |
| --- | --- | --- | --- |
| `method` | Yes | Connection method. Accepted values: `thrift`, `http`, or `livy`. | `thrift` |
| `port` | Yes | Port number for the connection. | `443` |
| `auth` | Yes | Authentication method. | `SASL PLAIN`, `NOSASL`, `AWS_SIGV4`  |
| `schema` | Yes | The database or schema name where dbt will create and query objects. | `analytics` |
| `host` | Yes | Hostname of the Spark cluster or Databricks workspace. | `yourorg.sparkhost.com` |

> | `platform_hint` | No | Hints to <Constant name="fusion" /> which Spark platform you use. Used to validate required `server_side_parameters`. Accepted values: `aws_emr_serverless`, `aws_emr_eks`. If omitted, Fusion assumes a generic Spark cluster. | `aws_emr_eks` |

Reviewer: ```suggestion
| `platform_hint` | No | Hints to <Constant name="fusion" /> which Spark platform you use. <Constant name="fusion" /> uses this to validate required `server_side_parameters`. Accepted values: `aws_emr_serverless`, `aws_emr_eks`. If omitted, <Constant name="fusion" /> assumes a generic Spark cluster. |
```

> | `server_side_parameters` | No | Spark session parameters passed to the cluster. <br/><br/>Required keys when using `platform_hint`:<br/>- For `aws_emr_serverless`, use `emr-serverless.session.executionRoleArn`.<br/>- For `aws_emr_eks`, use `spark.kubernetes.namespace`.<br/><br/>When using `method: livy`, you can set `livy.server.session.ttl` to configure how long a session can remain idle before it is terminated. | See [example profiles](#example-profiles). |

Reviewer: ```suggestion
| `server_side_parameters` | No | Spark session parameters passed to the cluster. <br/><br/>Required keys when using `platform_hint`:<br/>- For `aws_emr_serverless`, use `emr-serverless.session.executionRoleArn`.<br/>- For `aws_emr_eks`, use `spark.kubernetes.namespace`.<br/><br/>When using `method: livy`, you can set `livy.server.session.ttl` to configure how long a session can remain idle before it is terminated. | Refer to [example profiles](#example-profiles). |
```

## Authentication

The Spark adapter in <Constant name="fusion" /> supports:

- Thrift
  - Simple Authentication and Security Layer (SASL) PLAIN
  - No SASL (NOSASL) 
- Livy
  - Basic authentication (username and password)
  - When deployed on Amazon Web Services (AWS): AWS Signature Version 4
    - Supports authentication using single sign-on, service accounts, or user tokens

## Configure Fusion

Configure your Spark connection in `profiles.yml`:

<File name='~/.dbt/profiles.yml'>

```yaml
your_profile_name:
  target: dev
  outputs:
    dev:
      type: spark
      method: [thrift | http | livy]
      port: [port number]
      auth: [authentication method]
      schema: [database/schema name]
      host: [yourorg.sparkhost.com]
      platform_hint: [aws_emr_serverless | aws_emr_eks]  # optional
      server_side_parameters:  # optional; required keys depend on platform_hint
        "[key]": "[value]"

>       livy.server.session.ttl: [seconds]  # optional; when using method: livy

Reviewer 3: this is a key under `server_side_parameters`. Sorry I wasn't clear about it.

  config_page: '/reference/resource-configs/spark-configs'
---

<VersionBlock firstVersion="2.0">

# Connect Apache Spark to Fusion <Lifecycle status="beta" />

The <Constant name="fusion_engine" /> supports Apache Spark, enabling faster compilation and execution for your Spark-based dbt projects. Currently, <Constant name="fusion" /> only supports Apache Spark 3.0.

## Fusion and Spark

> <Constant name="fusion" /> uses the Databricks SQL dialect for [static analysis](/docs/fusion/new-concepts#principles-of-static-analysis) when working with Spark. This means your Spark SQL is validated using Databricks SQL semantics, providing comprehensive error checking and SQL comprehension features.

Reviewer 3: This is correct. The goal is to change to a dedicated Spark SQL frontend in the future though, since Databricks is a superset of Spark. Maybe we can say that's only a stopgap solution while we make a proper Spark dialect internally?

## Authentication

The Spark adapter in <Constant name="fusion" /> supports:

> - Service Account / User Token authentication

Reviewer 3: We're supporting:
- For Thrift: NOSASL or SASL PLAIN authentication. LDAP and KERBEROS will come later.
- For Livy: Basic (username/password authentication)
  - On AWS we also support AWS SigV4, which may work with SSO, service accounts or user tokens.

## Configure Fusion

Configure your Spark connection in `profiles.yml`:

<File name='~/.dbt/profiles.yml'>

```yaml
your_profile_name:
  target: dev
  outputs:
    dev:
      type: spark

>       method: odbc

Reviewer 3: Method `odbc` is not supported. We support
- `thrift`
- `http`
- `livy`

>       driver: [path/to/driver]

Reviewer 3: There is no `driver` path either.

      schema: [database/schema name]
      host: [yourorg.sparkhost.com]

>       token: [abc123]

Reviewer 3: `token` and `cluster` are Databricks-specific and don't exist either.

      cluster: [cluster id]
      platform_hint: aws_emr_serverless # or aws_emr_eks
```

</File>

> | Profile field | Required | Description | Example |

Reviewer 3: Examples of valid Spark profiles:

```yaml
spark-local-thrift-binary:
  target: spark
  outputs:
    spark:
      type: spark
      method: thrift 
      port: 10000
      auth: NOSASL
      host: localhost
      schema: my_schema

spark-local-thrift-http:
  target: spark
  outputs:
    spark:
      type: spark
      method: http
      port: 443
      auth: NOSASL
      # NOTE: not specifying the protocol scheme will assume HTTPS. If you need regular HTTP
      # use something like `http://localhost`
      host: localhost
      schema: my_schema

spark-emr-serverless:
  target: spark
  outputs:
    spark:
      type: spark
      method: livy
      auth: AWS_SIGV4
      port: 443
      host: "YOUR_APPLICATION_ID.livy.emr-serverless-services.us-east-1.amazonaws.com"
      schema: "my_schema"
      platform_hint: "aws_emr_serverless"
      server_side_parameters:
        "spark.driver.memory": "1g"
        "spark.driver.cores": 1
        "spark.executor.memory": "1g"
        "spark.executor.cores": 1
        # This property is required by EMR serverless. Adding `platform_hint: aws_emr_serverless`
        # will validate it's here
        "emr-serverless.session.executionRoleArn": "arn:aws:iam::YOUR_AWS_ACCOUNT:role/YOUR_ROLE"

spark-emr-eks:
  target: spark
  outputs:
    spark:
      type: spark
      method: livy
      auth: AWS_SIGV4
      port: 443
      host: "https://my-spark-cluster.com"
      schema: "my_schema"
      platform_hint: "aws_emr_eks"
      server_side_parameters:
        # This property is required by EMR on EKS. Adding `platform_hint: aws_emr_eks`
        # will validate it's here
        "spark.kubernetes.namespace": "emr-jobs"
        "spark.driver.memory": "1g"
        "spark.driver.cores": 1
        "spark.executor.memory": "8g"
        "spark.executor.cores": 4

```

Reviewer 3: I forgot to mention, users can set a `livy.server.session.ttl` parameter to configure the session TTL (i.e the time it can remain idle before being killed) when `method: livy`.
